Short read QC

jmtsuji opened the pull request:
This code adds basic short read QC functionality to rotary to address #38. Summary of changes:
- A new master checkpoint for QC was created in `rule qc`. This selects (on a per-run basis, for the MVP version) whether short + long read QC will be performed or just long read QC
- Short read QC starts with tossing out junk reads (e.g., with ambiguous bases) and checking file integrity with `rule short_read_reformat`
- Adapter removal (from the right side of reads) and quality trimming are then performed in the same step, `rule short_read_adapter_and_quality_trimming`. Quality trimming is performed after adapter removal so that adapters near the ends of reads can be more accurately identified (e.g., before some bases are clipped off the ends). I think ATLAS takes the same approach.
  - TODO: the adapters file from ATLAS is directly downloaded from Zenodo - do we want to make and host our own file?
  - Note that adapter trimming and quality trimming are currently non-optional. We can add the ability to toggle these on/off if you would like.
  - I set a relatively aggressive q-trim threshold of Q20
- Contaminant filtration is the last step, via `rule short_read_contamination_filter`. This step is optional based on a variable in the config.
  - I built in PhiX genome and human genome contaminant filtration into rotary. These genomes will be auto-downloaded into the DB folder if the user requests to screen with them.
  - The user can optionally add a single additional FastA nucleotide file as a screening reference.
  
I did not add good logging for QC stats. The user needs to manually look at log files to understand what happened during QC. In future, it could be nice to add a rule that parses the log files to summarize QC results.

Lastly, the current code passes `--dry-run` on my machine but has only been tested in smaller chunks with real data. A proper end-to-end test will be needed at some point (maybe once everything is merged into #71 ?).

@LeeBergstrand Mind adding your review? It looks like there will be no merge conflicts with improve_multi_sample_pathing (the branch that this branch was derived from), which is good news! Thanks.

LeeBergstrand commented on rotary/rules/rotary.smk:
```diff
@@ -8,9 +8,11 @@ import itertools
+from rotary.utils import symlink_or_compress, get_contamination_reference_files
 
 
+PHIX_GENOME_ACCESSION = "GCF_000819615.1"
```
@jmtsuji I think we should eventually move these versions to a separate configuration file. The database version list is starting to get quite long.

jmtsuji replied:
I agree -- I was thinking the same thing. Will convert to issue.

jmtsuji replied:
Done -- #78

LeeBergstrand commented on rotary/rules/rotary.smk:
```diff
@@ -36,13 +38,66 @@ else:
+        # Confirm that there is only one genome file matching the expected pattern in the unzipped folder
+        genome_file=($(find {output.zip_dir}/ncbi_dataset/data/{params.accession} -type f -name "{params.accession}_*_genomic.fna"))
+        if [[ "${{#genome_file[@]}}" == 1 ]]; then
+          gzip -c "${{genome_file[0]}}" > {output.genome}
```
If it's like 1-3 mb genome it's probably okay. I assume it won't take long.

jmtsuji replied:
The issue is the human genome -- it's ~1-3 GB (I was surprised by the size!). I tested it on my Mac, and the compression time with standard gzip is about 5-10 minutes, which is tolerable but a bit long. I've added pigz now to the rule (with threads capped at 4 max like the short_read_reformat rule below).

LeeBergstrand replied:
@jmtsuji My mistake; I forgot that this was the human genome, I was underreading. Yes, it makes a lot of sense to use pigz in this situation.

LeeBergstrand commented on rotary/rules/rotary.smk:
```diff
@@ -312,16 +367,180 @@ rule qc_long_length_stats:
+rule short_read_reformat:
+    """
+    Makes the input format of the short reads consistent (e.g., by trimming read descriptions and tossing invalid 
+    nucleotide characters). Threads are locked at a maximum of 4 because this code is IO limited.
```
@jmtsuji I think snakemake has a way to override resources for specific rules. What is interesting about this is that if you put this code on a RAM disk or potentially SSD you can use more threads. I think locking it at 4 threads is good for now.

LeeBergstrand replied:
I found that when I was running HMMER3 on ZFS data with the RAM cache I could use more threads than the 8 it normally bottlenecks at because the RAM cache was able to feed the algorithm fast enough to make a difference.

jmtsuji replied:
Good point that some hardware configurations could support higher thread counts. One good thing is that snakemake should be able to run multiple jobs in parallel that are below the total resources allotted to snakemake. So even if capping this job at 4 threads does not maximize the thread usage in all setups, at least other rules should be able to run simultaneously to fill the space. I agree, let's keep it at 4 threads for now, but we can always change this in future (or add another config variable somewhere, whether exposed or hidden, for the thread count of small thread jobs).

LeeBergstrand commented on rotary/rules/rotary.smk:
```diff
@@ -312,16 +367,180 @@ rule qc_long_length_stats:
+    input:
+        "{sample}/qc/long/{sample}_nanopore_qc.fastq.gz"
+    output:
+        "{sample}/qc/{sample}_QC_long.fastq.gz"
```
@jmtsuji This might be a bit nitpicky, but I was wondering if there was a reason why QC is capitalized? All the other files with QC in their file names have the 'qc' lower case.

jmtsuji replied:
Thanks for noticing this! I've converted all to lowercase.

LeeBergstrand commented on rotary/utils.py:
```diff
@@ -100,3 +100,39 @@ def symlink_or_compress(in_file_path, out_file_path):
+    contamination_reference_paths = []
+
+    if isinstance(config_entry, list) is True:
+
```
@jmtsuji I haven't brought this up before, but in some situations, you can drop the blank lines below the for and if statements and let the indentation handle the readability.

```python
if isinstance(config_entry, list) is True:
        for contamination_entry in config_entry:
            if contamination_entry.lower() == 'phix':
                contamination_reference_paths.append(os.path.join(db_path, 'contamination_references',
                                                                  f'PhiX__{phix_accession}.fna.gz'))
            elif contamination_entry.lower() == 'human':
                contamination_reference_paths.append(os.path.join(db_path, 'contamination_references',
                                                                  f'human__{human_accession}.fna.gz'))
            else:
                raise ValueError(f'Contaminant genome must be PhiX or human, but you provided {contamination_entry}')
else:
    raise TypeError(f'Contaminant genome names must be in a list, but you provided type {type(config_entry)}. '
                    'Try editing the config file and make sure that contamination_references is a list like '
                    '[PhiX] or [PhiX, human], not just "PhiX" or "human".')
```

LeeBergstrand replied:
I usually use blank lines to break out logical blocks. For example:

```python
if x: # no space here because it's redundant because of the indent.
   data=some_func()
           # space here to break out the for loop away from the other code. Since blam is filled
   blam=[] # by the loop, we put it directly above the loop with no space and break it out.  
   for datum in data:
       if datum.cool == True: # Some libraries like numpy they have their own boolean values  
          blam.append(datum)  # (```np.bool```) so you have to use == to force an evaluation. 

   if len(blam) > 1: # Blank line between for loop and if statement since they are different logics.
       print(blam)
```

LeeBergstrand replied:
@jmtsuji Just some examples of how I use spacing.

LeeBergstrand commented on rotary/utils.py:
```diff
@@ -100,3 +100,39 @@ def symlink_or_compress(in_file_path, out_file_path):
+
+    contamination_reference_paths = []
+
+    if isinstance(config_entry, list) is True:
```
Also, you can remove the is True here and it should still work.:

```python
isinstance(config_entry, list):
   do stuff..
```

I usually do this with functions that evaluate something for example functions that evaluate only true or false. I.e., is the config entry an instance of list? Yes or No.

jmtsuji replied:
Good to know I can simplify this -- I was trying to keep the True/False test obvious in the code for readability, but I think you're right that the function name itself makes it pretty clear what the code is doing. Just addressed this.

LeeBergstrand commented:
@jmtsuji This looks good so far. I left a few comments.

jmtsuji commented:
@LeeBergstrand Addressed your comments -- mind taking another look?

Also, what are your thoughts on the following (pasted from the PR description above)?
> - TODO: the adapters file from ATLAS is directly downloaded from Zenodo - do we want to make and host our own file?
> - Note that adapter trimming and quality trimming are currently non-optional. We can add the ability to toggle these on/off if you would like.
> - I set a relatively aggressive q-trim threshold of Q20 [as default]

LeeBergstrand commented:
@jmtsuji 

> > * TODO: the adapters file from ATLAS is directly downloaded from Zenodo - do we want to make and host our own file?

We could put this down as a long-term TODO. I wonder if the archive will go down or if ATLAS will stop using this URL.

> > * Note that adapter trimming and quality trimming are currently non-optional. We can add the ability to toggle these on/off if you would like.

This should be implemented. However, I would like to merge the QC code into develop beforehand.

> > * I set a relatively aggressive q-trim threshold of Q20 [as default]

This is fine for Illumina, which should have reasonable error rates. We should have more stringent cutoffs since we are polishing with the short reads.

jmtsuji commented:
@LeeBergstrand 
> We could put this down as a long-term TODO. I wonder if the archive will go down or if ATLAS will stop using this URL.

Addressed - the code now pulls from a rotary-specific data repo that contains the adapters file: https://zenodo.org/records/10087395
Attribution is given to the ATLAS repo.

jmtsuji commented:
> This should be implemented. However, I would like to merge the QC code into develop beforehand.

OK, sounds good.

> This is fine for Illumina, which should have reasonable error rates. We should have more stringent cutoffs since we are polishing with the short reads.

Great, let's proceed with Q20 for now, then.

jmtsuji commented:
> > Note that adapter trimming and quality trimming are currently non-optional. We can add the ability to toggle these on/off if you would like.

> This should be implemented. However, I would like to merge the QC code into develop beforehand.

@LeeBergstrand FYI, I had a bit of extra time, so I went ahead and added the ability to toggle adapter and quality trimming on/off into the code. Basically:
- I had to split adapter trimming and quality trimming into 2 separate rules
- The adapter trimming rule does not run if the config flag `perform_adapter_trimming` is set to False
- The quality trimming rule always runs, but if `quality_trim_direction` is set to "f" in the config (an actual setting in bbduk), then the reads are just passed through the rule without actual trimming. This simplified the logic in the snakefile quite a bit.

LeeBergstrand commented on rotary/rules/rotary.smk:
```diff
@@ -312,16 +369,228 @@ rule qc_long_length_stats:
+    params:
+        contamination_filter_kmer_length = config.get("contamination_filter_kmer_length"),
+        all_contaminant_references = ','.join(get_contamination_reference_files(config.get("contamination_references"),
+                DB_DIR_PATH, PHIX_GENOME_ACCESSION, HUMAN_GENOME_ACCESSION)) \
```
@jmtsuji Move this to a function passed to params (I think that's possible?), or use this code to assign a global variable at the top of the file. This is too long of an inline if-else statement. Once the code is moved above, you can convert it to a traditional if-else statement using pycharm (right-click the if and select show context actions ). Also, you could assign the result of `get_contamination_reference_files(config.get("contamination_references"), DB_DIR_PATH, PHIX_GENOME_ACCESSION, HUMAN_GENOME_ACCESSION)` to the second global variable, which could be used in the prior mentioned if-else statement and could also be assigned to ```input.contamination_references```  for rule ```short_read_contamination_filter```

jmtsuji replied:
@LeeBergstrand This variable was simplified so that an if-else statement is no longer needed. It's still a bit long, though (mostly because of long variable and function names) -- what do you think?

jmtsuji replied:
Also, note that this will fail with a somewhat cryptic error message if the user provides bad input (e.g., a string, not a list) to `contamination_references_custom_filepaths` in the config. We could add a sanity check for this in the config check proposed in #79 . Otherwise, I could make a function to validate these input files.

LeeBergstrand commented on rotary/rules/rotary.smk:
```diff
@@ -312,16 +369,228 @@ rule qc_long_length_stats:
+    input:
+        short_r1 = "{sample}/qc/short/{sample}_quality_trim_R1.fastq.gz",
+        short_r2 = "{sample}/qc/short/{sample}_quality_trim_R2.fastq.gz",
+        contamination_references = get_contamination_reference_files(config.get("contamination_references"),
```
move to global variable.

jmtsuji replied:
Done

LeeBergstrand commented on rotary/utils.py:
```diff
@@ -100,3 +100,34 @@ def symlink_or_compress(in_file_path, out_file_path):
         gzip_file(in_file_path, out_file_path)
+
+
+def get_contamination_reference_files(config_entry: list, db_path: str, phix_accession: str, human_accession: str):
```
Change the name of this function to `get_contamination_reference_file_paths.` Generally, when I put the word `file` of a function, I expect it to return a file handle, which is the object created by the `open('file/path).` This function returns the paths to the files, not the files or handles themselves.

jmtsuji replied:
Done - good point!

LeeBergstrand commented:
Improve by putting contaminant NCBI IDs in the config and also a list of paths to custom contaminant FASTA. As discussed below:

![image](https://github.com/jmtsuji/rotary/assets/5819462/2b25ec75-f0f3-45d0-80c3-37f37d84b941)

jmtsuji commented:
> Improve by putting contaminant NCBI IDs in the config and also a list of paths to custom contaminant FASTA.

Done! Let me know your thoughts on the updated code.

Aside from addressing the core changes you requested, some other things were also added:
- warning was added in the config file about ~60 GB memory needed for human genome-based decontamination
- pigz / unpigz is also forced on now

LeeBergstrand commented on rotary/config.yaml:
```diff
@@ -15,6 +15,43 @@ minlength: 1000
+minimum_read_length_quality_trim: 10
+## Contaminant filtration
+# Search for and remove reads that match contaminant reference sequences?
+perform_contaminant_filtration: "True"
```
@jmtsuji What are your thoughts about removing this parameter and having it so that the sets both `contamination_references_ncbi_accessions` and `contamination_references_custom_filepaths` to `[]` to turn off the contamination filtering. Having this `perform_contaminant_filtration` flag is redundant now with your recent changes. What are your thoughts?

LeeBergstrand replied:
Implement in https://github.com/jmtsuji/rotary/pull/82

LeeBergstrand commented on rotary/rules/rotary.smk:
```diff
@@ -312,16 +369,222 @@ rule qc_long_length_stats:
+    input:
+        short_r1 = "{sample}/qc/short/{sample}_quality_trim_R1.fastq.gz",
+        short_r2 = "{sample}/qc/short/{sample}_quality_trim_R2.fastq.gz",
+        contamination_references = CONTAMINANT_REFERENCE_GENOMES_NCBI
```
@jmtsuji Will this fail if CONTAMINANT_REFERENCE_GENOMES_NCBI = None? How will this work if there are no NCBI contamination references?

LeeBergstrand replied:
Addressed in https://github.com/jmtsuji/rotary/pull/82

LeeBergstrand commented on rotary/rules/rotary.smk:
```diff
@@ -312,16 +369,222 @@ rule qc_long_length_stats:
+          2> {log}
+        """
+
+
```
Shouldn't this be in an if block?

LeeBergstrand replied:
Addressed in https://github.com/jmtsuji/rotary/pull/82

LeeBergstrand commented on rotary/rules/rotary.smk:
```diff
@@ -312,16 +369,222 @@ rule qc_long_length_stats:
+        "{sample}/benchmarks/qc/short/short_read_contamination_filter.benchmark.txt"
+    params:
+        contamination_filter_kmer_length = config.get("contamination_filter_kmer_length"),
+        all_contaminant_references = ','.join(CONTAMINANT_REFERENCE_GENOMES_NCBI +
```
I would move this logic to the top and assign it to a variable called "ALL_CONTAMINANT_REFERENCES." If contaminant references are none, then skip the contaminant filtering rule altogether.

LeeBergstrand replied:
Addressed in https://github.com/jmtsuji/rotary/pull/82

LeeBergstrand reviewed (changes requested):
Thanks. It looks like we are very close to finishing this branch. Good Job I highlighted a few more changes.

LeeBergstrand commented:
@jmtsuji I'm working on an elegant solution to the short-read contaminant set up.

jmtsuji commented:
@LeeBergstrand Thanks for the proposed revision! I just added my comments at #82

jmtsuji commented:
@LeeBergstrand I think this PR is only waiting on #82, correct? If so, then feel free to merge this PR after finishing the #82 merge. Otherwise, feel free to keep this PR open if you'd like to do another code review.
